jax/interpreters/xla.py

```diff
@@ -56,6 +56,148 @@ def _map(f, *xs): return tuple(map(f, *xs))
 def identity(x): return x
 
 
+### lazy sub-language
+
+LazyExpr = namedtuple('LazyExpr', ['input', 'shape', 'dims'])
+LazyArrayVar = namedtuple('ArrayVar', [])
+LazyIota = namedtuple('Iota', ['dtype', 'size'])
+LazyEye = namedtuple('Eye', ['dtype', 'shape', 'offset'])
+LazyTri = namedtuple('Tri', ['dtype', 'shape', 'offset'])
```

Comment:
Eye and Tri model exactly np.eye and np.tri. Delta models Kronecker delta arrays.

I think this is a borderline case for documentation: it's easy for an interested party to figure out exactly what these things do by looking at the code. I'll add very short descriptions.

I'm not going to add a TODO because we could apply a "TODO: always be improving" everywhere in the code. I would add a TODO if we had a specific plan to make these better, but we don't. None of the more generic mechanisms I've prototyped or thought about would be a good fit here (because they'd be heavyweight, or would require making jaxprs hashable, etc).

Reply:
Done.

Reply:
I thought about it more, and what I said was wrong. This is a good case for documentation. Sorry.

I still think the same about the TODO though :)

Reply:
I added I added this sentence to the main comment in this section:

> For the array constructor expressions, Iota builds a 1D sequence [0, 1, ..., N-1], Eye builds a 2D array with ones on a (possibly offset) diagonal and zeros elsewhere (like numpy.eye), Tri builds a triangular matrix with ones on and below a diagonal and zeros elsewhere (like numpy.tri), and Delta builds a Kronecker delta array with ones along its multidimensional main diagonal and zeros elsewhere.

I also added same-line comments on the namedtuple definition lines.

WDYT?